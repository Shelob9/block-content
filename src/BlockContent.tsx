import { Block, parse } from '@wordpress/block-serialization-default-parser';
import React, { createElement, ReactNode, useMemo, FC } from 'react';
import { addAndParseBlock } from './transformNodes';

/**
 * Render block content with fallback to rendered.
 *
 * Safer option than default export
 */
export const RenderBlockContent: FC<{ raw?: string; rendered: string }> = ({
  raw,
  rendered,
}) => {
  function createMarkup() {
    return { __html: rendered };
  }
  if (!raw) {
    return <div dangerouslySetInnerHTML={createMarkup()} />;
  }
  return <BlockContent rawContent={raw} />;
};
export type PROPS = {
  rawContent: string;
};
/**
 * Render a block
 *
 * Only works with raw block content.
 */
const BlockContent = (props: PROPS) => {
  let { rawContent } = props;
  let Component = useMemo(() => {
    let blocks = parse(rawContent);
    let els: ReactNode[] = [];
    Object.values(blocks).forEach((block: Block) => {
      addAndParseBlock(block, els);
    });
    return () => createElement('div', {}, els);
  }, [rawContent]);

  return (
    <>
      <Component />
    </>
  );
};

export default BlockContent;
