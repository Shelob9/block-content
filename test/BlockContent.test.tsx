import React from 'react';
import { render } from '@testing-library/react';
import { paragraphs } from '../fixtures/blockContents';
import BlockContent, {transform} from '../src/index';

describe('BlockContent', () => {
  it('renders', () => {
    const { container} = render(<BlockContent rawContent={paragraphs} transformer={transform}  />);
    expect(container).toMatchSnapshot();
  });

  it('Prints a paragraph', () => {
    const { getByText} = render(<BlockContent rawContent={`<!-- wp:paragraph -->
    <p>Hi Roy</p>
    <!-- /wp:paragraph -->`} transformer={transform}  />);
    expect(getByText('Hi Roy')).toEqual('<p>Hi Roy</p>');
  });
});
