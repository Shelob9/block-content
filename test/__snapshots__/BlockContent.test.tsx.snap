// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockContent Prints a paragraph 1`] = `
<div>
  <div>
    <p
      class="noms"
    >
      <em>
        First
      </em>
    </p>
    <p>
      <em>
        Second
      </em>
    </p>
  </div>
</div>
`;

exports[`BlockContent renders 1`] = `
<div>
  <div>
    <p
      class="noms"
    >
      <em>
        First
      </em>
    </p>
    <p>
      <em>
        Second
      </em>
    </p>
  </div>
</div>
`;

exports[`BlockContent renders without crashing 1`] = `
<div>
  <div>
    <p
      class="noms"
    >
      <em>
        First
      </em>
    </p>
    <p>
      <em>
        Second
      </em>
    </p>
  </div>
</div>
`;

exports[`Thing renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p
          class="pretzels"
        >
          <em>
            First
          </em>
        </p>
        <p
          class="pretzels"
        >
          <em>
            Second
          </em>
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p
        class="pretzels"
      >
        <em>
          First
        </em>
      </p>
      <p
        class="pretzels"
      >
        <em>
          Second
        </em>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
